### ISSUE TITLE
Error: TIMEOUT: Port 8080 on localhost was not active within 30000ms

### ISSUE THREAD

**Issue by dave-k** [firebase/firebase-tools#2379]
[firebase-debug.log](https://github.com/firebase/quickstart-nodejs/files/4795530/firebase-debug.log)
Cloud Firestore emulator quickstart
I am getting the Timeout error running the ````emulators:exec```` command on an unchanged project.

I tried other ports ````8081, 8082```` and 

````
$ sudo firebase emulators:exec --only firestore 'npm test'
$ firebase emulators:exec 'npm test'
$ firebase emulators:start
```` 
produced the same Timeout error.

````

./firebase/quickstart-nodejs/firestore-emulator/javascript-quickstart

macOS v 10.15.6

$ firebase --version
8.4.2
$ node --version
v14.4.0
$ 

````

````
$ firebase emulators:exec --only firestore 'npm test'
i  emulators: Starting emulators: firestore
i  firestore: Firestore Emulator logging to firestore-debug.log
i  emulators: Shutting down emulators.

Error: TIMEOUT: Port 8080 on localhost was not active within 30000ms
````

**Comment by dave-k** [comment]
@samtstern Unchanged i.e. no changes made to the tests.
1. node v10.21.0
2. host: 0.0.0.0

Same Timeout error: [error] Error: TIMEOUT: Port 8080 on 0.0.0.0 was not active within 30000ms
[firebase-debug.log](https://github.com/firebase/firebase-tools/files/4807268/firebase-debug.log)

firebase.json
````
{
  "firestore": {
    "rules": "firestore.rules"
  },
  "emulators": {
    "firestore": {
      "host": "0.0.0.0",
      "port": 8080
    }
  }
}
````

package.json
````
{
  "name": "cloud-firestore-emulator-javascript-quickstart",
  "version": "1.0.2",
  "description": "Cloud Firestore emulator testing",
  "scripts": {
    "format": "prettier --write **/*.js",
    "test": "mocha --timeout=10000"
  },
  "devDependencies": {
    "@firebase/testing": "^0.20.3",
    "mocha": "5.2.0",
    "prettier": "1.15.2"
  }
}
````

**Comment by samtstern** [comment]
@dave-k thanks for trying hose things ... I am still pretty stumped on this issue.

Could you try running the Firestore emulator JAR directly and see if it works for you:
```
java -jar ~/.cache/firebase/emulators/cloud-firestore-emulator-v1.11.4.jar
```

The exact version of the `jar` may vary but you should have a `cloud-firestore-emulator-*.jar` in that location.  You can also pass `--port=<number>` if you don't want to use `8080`.

**Comment by dave-k** [comment]
@samtstern 

````
$ java -jar ~/.cache/firebase/emulators/cloud-firestore-emulator-v1.11.4.jar
API endpoint: http://localhost:8080
If you are using a library that supports the FIRESTORE_EMULATOR_HOST environment variable, run:

   export FIRESTORE_EMULATOR_HOST=localhost:8080

Dev App Server is now running.
````

**Comment by dave-k** [comment]
````
$ java -Duser.language=en \
>   -jar /Users/dkennedy//.cache/firebase/emulators/cloud-firestore-emulator-v1.11.4.jar \
>   --host "0.0.0.0" --port 8091 \
>   --rules /Users/dkennedy/project-2/firebase/quickstart-nodejs/firestore-emulator/javascript-quickstart/firestore.rules
API endpoint: http://0.0.0.0:8091
If you are using a library that supports the FIRESTORE_EMULATOR_HOST environment variable, run:

   export FIRESTORE_EMULATOR_HOST=0.0.0.0:8091

Dev App Server is now running.
````

**Comment by dinvlad** [comment]
We have the same issue with `cloud-firestore-emulator-v1.11.2.jar`

**Comment by Ross-Rawlins** [comment]
any news on this? I am getting this issue as well when importing a large data set.

**Comment by samtstern** [comment]
@Ross-Rawlins that sounds like a different issue since nobody above has mentioned import!  For the issue above we have not been able to reproduce it ourselves, and we really can't fix an issue we can't reproduce (although we acknowledge that it's a real issue)

**Comment by beezital** [comment]
Same issue on my side running MacOS too.
It seems to be related to a networking condition involving `java`.
I have timed `java -jar ~/.cache/firebase/emulators/cloud-firestore-emulator-v1.11.5.jar` to be 31 seconds before it succeeds starting. Obviously this is too long when `firebase emulators:start` times out after 30 seconds.
I have a multi-WAN router establishing multiple network connections on different WANs (works great for standard browsing where fetching can happen in parallel). I have experienced other issues with this setup and involving `java`.
**Workaround**
So I tunneled my Mac's Internet access through a VPN and now `java -jar ~/.cache/firebase/emulators/cloud-firestore-emulator-v1.11.5.jar` starts instantaneously. And so does `firebase emulators:start`.

**Comment by handersonc** [comment]
Same issue here when I run `firebase emulators:start --only firestore --import=./myfolder`

**Comment by dracomithril** [comment]
So I was fighting with that thing for some time checking every possibility and the I did 
```
npm install -g firebase-tools
``` 

and surprise surprise it started working  🤷 why bits me but maybe it will help someone

**Comment by ScottPierce** [comment]
I'm having this same problem in Docker Ubunbu environments trying to run CI tests. This morning, it just kept failing left and right for some reason:

```
yarn run v1.22.5
$ cd firebase && firebase emulators:exec --only firestore "cd .. && yarn test"
⚠  emulators: You are not currently authenticated so some features may not work correctly. Please run firebase login to authenticate the CLI.
i  emulators: Starting emulators: firestore
i  firestore: Firestore Emulator logging to firestore-debug.log
i  emulators: Shutting down emulators.
i  firestore: Stopping Firestore Emulator
i  hub: Stopping emulator hub

Error: TIMEOUT: Port 8080 on 0.0.0.0 was not active within 30000ms
```

**Comment by dracomithril** [comment]
Did you add FIREBASE_TOKEN as env variable as it's mentioned in [documentation ](https://firebase.google.com/docs/cli#cli-ci-systems)

**Comment by ScottPierce** [comment]
@dracomithril I don't think that's necessary to run tests. My current theory is that the computer that was running the build was constrained on resources, creating a slow startup time. We started seeing tests time out also. I reduced the number of parallel builds running on the server at one time, and haven't seen it since.

**Comment by dracomithril** [comment]
one probably odd question why
```
cd firebase && firebase emulators:exec --only firestore "cd .. && yarn test"
```
and not 
```
"scripts": {
    ...
    "test": "firebase emulators:exec --only firestore jest"
  },
```

and regarding FIREBASE_TOKEN 
` You are not currently authenticated so some features may not work correctly. Please run firebase login to authenticate the CLI.`
in my case it helped 🤷

**Comment by ScottPierce** [comment]
> one probably odd question why

Because `yarn test` is the following script:
```
nx run-many --all --target=test --parallel --maxParallel 3
```

The above command is fairly complicated, and instead of duplicating it in multiple places, I'd rather centralize it into one place.

**Comment by brennengarland** [comment]
I've been having this same problem the last few days out of the blue. I'm running `firebase emulators:start --import=data` and the output is:

```
i  emulators: Starting emulators: auth, functions, firestore, database, hosting
!  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: pubsub
!  Your requested "node" version "10" doesn't match your global version "12"
i  firestore: Importing data from C:\repos\proctr\data\firestore_export\firestore_export.overall_export_metadata
i  firestore: Firestore Emulator logging to firestore-debug.log
i  emulators: Shutting down emulators.
i  functions: Stopping Functions Emulator
i  firestore: Stopping Firestore Emulator
!  Firestore Emulator has exited upon receiving signal: SIGINT
i  hub: Stopping emulator hub

Error: TIMEOUT: Port 8080 on localhost was not active within 30000ms
```
This problem persists if I change the ports.

**Comment by dracomithril** [comment]
@brennengarland did you changed node version?
```
! Your requested "node" version "10" doesn't match your global version "12"
```
If yes maybe try to reinstall all packages 🤔

**Comment by ScottPierce** [comment]
@brennengarland If you are running it on CI, then you might not have a fast enough CI container. Otherwise, I'd be looking at how large your data import is. If the import happens before the emulator is live, that could be why it's taking longer than 30 seconds.

**Comment by brennengarland** [comment]
@ScottPierce it looks like independent of the data because when I run `firebase emulators:start --only firestore` I have the same issue. It might just be my machine, but I have no idea what would be causing the ports to be taking so long to open

**Comment by samhorlbeck** [comment]
Hi folks, this definitely sounds like a bug (tracked internally at b/172852315).

**Comment by nolafs** [comment]
firebase emulators:start --only firestore works for me fine but when try to import the database, I'm getting the Timeout error. The database is large, my guess is that import is taking to long.

**Comment by evilandfox** [comment]
I had the same problem, but it resolved after adding `rules_version = '2';` line on the beginning of `firestore.rules` file

**Comment by kevin-mao** [comment]
Having same issue as @nolafs - works fine without importing but shows this Timeout error when importing a rather large database (import folder is 460.3 MiB). Is there any way to extend timeout?

**Comment by goelrishabh** [comment]
After trying all the methods/steps mentioned above none of them worked for me BUT _I found the solution on my machine so leaving it here as a reference_.

The issue is caused by the multiple Network Adapters created by **Oracle VirtualBox VM**, I found that:

1. Uninstalling the **Oracle VirtualBox VM** removes these Virtual Host adapters and solves the **Timeout Error** issue
2. Disabling these Virtual Host adapters also solves the **Timeout Error** issue

I don't know exactly the reason why Disabling/Removing these Virtual Host adapters are solving the issue and the only possible statement that I can think of is that **Firebase Emulator has a limit of 30000ms after which it gets Timeout but running the jar file directly as stated by @samtstern opens the Dev server but takes a lot of time, This time is due to the presence of those Virtual Host adapters because disabling them individually one by one was indeed increasing the time and finally there is a threshold point when 30000ms is not the sufficient time window to (say)SCAN all of them!**

**Comment by mrenrich84** [comment]
In my case, I have to start it with the laptop (MacOS Big Sur 11.3) in airplane mode.
In this way, it possibly fails -  but then finally work on the next launch, also with the wifi back on.

I noticed that if I manually launch the jar before doing the above trick, it will take ages to launch the firestore server.
Disabling the wifi takes the startup to few secs.

I don't have any Virtual Adapter but am using the Cisco Umbrella Roaming Client - a DNS enterprise client that gives more than few headaches on my machine

**Comment by mrenrich84** [comment]
I think I found the solution: Firestore takes more than 30 secs to start.
The timeout value is hardcoded so as users we can't do anything really.

I opened #3482 to:
- increase the default timeout to 60 secs
- leave open to the user the configuration of the timeout for each emulator in the same pattern of host and port

**Comment by mrenrich84** [comment]
Closing #3482 and opening one PR just for hardcoding the timeout to 60 secs.

**Comment by Yoyoyoyoyoyoyo** [comment]
Now I'm getting timeouts at 60 seconds instead of 30.

**Comment by jdrin** [comment]
Experiencing the same issue - timeouts with 60000ms, calling jar directly works and all other methods time out.

**Comment by huydinhle** [comment]
Both downgrading and adding the host don't work for me but I'm trying to import large data set ( about 40gb). So there is no workarond for importing large data set?

**Comment by bjornconstructors** [comment]
For people running `firebase emulators:start` with the `--import` flag on large data sets like @huydinhle and @nolafs: my suggestion would be to _not_ use `--import` but store whatever data store you have in a JSON file and import that _after_ the emulators are running using a custom made population script. Your development flow would then to simply run this population script (each time) after starting the emulator. That way the importing process can take as long as you want and you could even add a nice progress bar as a bonus ;). Note that you likely also need a small script to dump the firestore data to that JSON file in the first place. Other only other workaround there is to fork firebase-tools and adapt the timeout in the source code here https://github.com/firebase/firebase-tools/blob/0be4074d717c0e58f36ffe5b0cdc09cffe869f1f/src/emulator/portUtils.ts#L142 .

**Comment by huydinhle** [comment]
@bjornconstructors : Have 2 follow up questions for you about this
1. How can you export your whole Firestore dataset into JSON ? Currently, I just use the export functionality from gcloud cli to get a backup from Firestore. How do I convert that 44 GB of data into JSON file? Is this going to be multiple JSON files?
2. Can you give us some pointer in terms of how to create this population script? You mentioned about progress bar even so You have some thing in mind in how to make this work?

**Comment by benedelstein** [comment]
Specifying a host in firebase.json seemed to work for me, but then my firestore-triggered cloud functions were not firing, with the error 
`WARNING: Failed to publish event to localhost:5001: Connection refused`

I downgraded to node 16, removed the host key, and now it works.

**Comment by TPXP** [comment]
Hello there, I'm having a similar issue on MacOS Monterey with Node v17.
Adding the `host` key to the configuration file fixed the startup, and gave me a possibility for the root cause of the issue. Indeed, I had the following line at the end of the my tests (to generate a firestore rules coverage report) 

```javascript
http.get(`http://127.0.0.1:${port}/emulator/v1/projects/${projectId}:ruleCoverage.html`, res => {
 // ...
})
```

This seems to crash as well ! Here's what I got
```
Error: connect ECONNREFUSED ::1:8090
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1161:16)
Emitted 'error' event on ClientRequest instance at:
    at Socket.socketErrorListener (node:_http_client:447:9)
    at Socket.emit (node:events:390:28)
    at emitErrorNT (node:internal/streams/destroy:164:8)
    at emitErrorCloseNT (node:internal/streams/destroy:129:3)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '::1',
  port: 8090
}
```

Replacing `localhost` by `127.0.0.1` fixed it, but I think `localhost` now resolves to `::1`(IPv6) instead of `127.0.0.1` (IPv4). Maybe that's why the CLI thinks the emulator is not yet ready

**Comment by stansotn** [comment]
I tried numerous things mentioned above. The only working solution for me was to downgrade to node@16 (from 17). I'm on Monterey 12.0.1 x86.

```shell
brew install node@16

# Relink node 17 installation to node 16
ln -sf /usr/local/Cellar/node@16/16.13.0/bin/node /usr/local/bin/node
```

**Comment by clouddevil** [comment]
As @stansotn mentioned,  downgrade to 16 node work for me too.  Catalina (10.15.7), tools - 9.22.0
```
brew unlink node
brew install node@16
brew link --overwrite node@16
```

**Comment by jangruenwaldt** [comment]
> As @stansotn mentioned, downgrade to 16 node work for me too. Catalina (10.15.7), tools - 9.22.0
> 
> ```
> brew unlink node
> brew install node@16
> brew link --overwrite node@16
> ```

Can confirm, working on monterey. Thanks for the commands.

**Comment by ruohola** [comment]
> I got the emulators running again by specifying host as "127.0.0.1" instead of the default "localhost". I have no idea what changed that caused it to stop working on localhost. All I did was modify firebase.json like so to add `host`
> 
> ```
>  "emulators": {
>         "auth": {
>             "port": 9099,
>             "host": "127.0.0.1"
>         },
>         "functions": {
>             "port": 5001,
>             "host": "127.0.0.1"
>         },
>         "firestore": {
>             "port": 8080,
>             "host": "127.0.0.1"
>         },
>         "hosting": {
>             "port": 5001
>         },
>         "ui": {
>             "enabled": true
>         },
>         "pubsub": {
>             "port": 8085,
>             "host": "127.0.0.1"
>         },
>         "storage": {
>             "port": 9199,
>             "host": "127.0.0.1"
>         }
>     }
> ```

Worked for me, I also needed to change all my env variables to not use `localhost`: `process.env.FIREBASE_DATABASE_EMULATOR_HOST = '127.0.0.1:9000';`

**Comment by supremeabraxas** [comment]
Downgrading to 16 resolved the issue for me.
Arch Linux (for search engines)

**Comment by yuchenshi** [comment]
Closing since downgrading to Node.js 16 seems to be a good workaround. Feel free to comment if you have more information or if this problem still persists when Node.js 18 becomes the new stable release.

**Comment by msrumon** [comment]
Confirmed that https://github.com/firebase/firebase-tools/issues/2379#issuecomment-951884721 works. I'm on Windows 10 Pro machine with Node.js v17.2.0.

**Comment by svenjacobs** [comment]
I'm having the same timeout issue with Node 17. In case you are using [Volta](https://volta.sh) for managing your Node versions, you can use the `volta` command for running `firebase` with a specific Node version, for example:

```
volta run --node 16.14.0 firebase emulators:start
```

**Comment by ninest** [comment]
I found this issue after Googling the error

**Comment by ktsin** [comment]
Just FYI same behavior on Windows machine with latest firebase tool
```
node -v
v18.5.0
```

```
firebase --version
11.2.2
```

Adding `"host" : "127.0.0.1"` resolved my issue.

**Comment by bkendall** [comment]
(internal bug reference: b/172852315)

**Comment by christhompsongoogle** [comment]
Hi everyone, there have been some changes recently to the IP resolution logic - if anyone is still hitting this issue please leave a comment.

**Comment by VikashFluxon** [comment]
I faced similar error where it failed to initialise the ports and faced a timeout while checking for it to be active.
I checked the logs and it had the following message. 
'auth: Port 5002 is available on 127.0.0.1 but not ::1. This may cause issues with some clients.'
I suggested me to change the port and try. I did and it worked but I am still curious as to why it did not work with 500X ports.

This started happening after the latest firebase release.
https://github.com/firebase/firebase-tools/releases/tag/v11.14.4

**Comment by christhompsongoogle** [comment]
Hi VikashFluxon, please open a new ticket with the relevant repro case(s) and logs.

**Comment by christhompsongoogle** [comment]
Is the problem reproducible with smaller sets of data? < 100MB type of size

**Comment by HandsonOmbiro** [comment]
Is there a step by step known documentation on how to configure firebase emulator to work with either the android/ios emulator or read devices, for me, NONE of the above has ever worked for me. The only thing which works is using the firebase emulators with web (Chrome or Edge) or connecting directly to Firebase live. Those works like charm. But connecting android emulator with firebase emulators or real device with firebase emulator has NEVER worked despite all those snip steps.  Is there a known step by step procedure on this, this will be useful. The documentation on Firebase seems to miss a lot, and even the codelabs still do not point exactly what is to be done when, where and how. To me it seems it is all guesswork.
